javascript/ql/src/semmle/javascript/dataflow/DataFlow.qll

```diff
@@ -1299,6 +1295,44 @@ module DataFlow {
     )
   }
 
+  /**
+   * Flow steps shared between `getImmediatePredecessor` and `localFlowStep`.
+   *
+   * Inlining is forced because the two relations are indexed differently.
+   */
+  pragma[inline]
+  private predicate immediateFlowStep(Node pred, Node succ) {
+    exists(SsaVariable v |
+      pred = TSsaDefNode(v.getDefinition()) and
+      succ = valueNode(v.getAUse())
+    )
+    or
+    exists(Expr predExpr, Expr succExpr |
+      pred = valueNode(predExpr) and succ = valueNode(succExpr)
+    |
+      predExpr = succExpr.(ParExpr).getExpression()
```

Comment:
Does it make the optimizer unhappy if we use `Expr::getUnderlyingValue` instead of all these explicit cases? (I know this is just a port of some steps, feel free to ignore)

Reply:
I don't know about the optimizer, but it doesn't do quite the same thing. `getUnderlyingValue` goes all the way to the bottom in one step, whereas we generally want the individual steps here.

For example, suppose you add type casts as a sanitizer in a data flow query. In this case, `getUnderlyingValue` would simply skip over the type cast, bypassing the sanitizer.

Reply:
Good point. I suppose we could look into letting `Expr` expose the non-recursive version as `getImmediatelyUnderlyingValue` in another PR.